File: q9-cycle-detection/q9-brute.js
// LeetCode 142.
// https://leetcode.com/problems/linked-list-cycle-ii/

const detectCycle = function (head) {
    if (!head)
        return null;
    
    let cur = head;
    const seenNodes = new Set();
    
    while (!seenNodes.has(cur)) {
        if (!cur.next)
            return null;
        
        seenNodes.add(cur);
        cur = cur.next;
    }
    
    return cur;
};

// time: O(n)
// space: O(n)

File: ztm-ds-algo/3-linked-list/q9-cycle-detection/python/q9-brute.py
# // LeetCode 142.
# // https://leetcode.com/problems/linked-list-cycle-ii/

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head):
        
        while head is not None:
            if head.val is None:
                return head
            head.val = None
            head = head.next
        return None
    
# time: O(n)
# space: O(1)